turbinia/lib/libcloudforensics.py

```diff
@@ -437,6 +437,66 @@ def get_or_create_analysis_vm(
     created = True
     return instance, created
 
+  def list_instance_by_label(self, labels_filter=''):
+    """Lists VMs in a project with on of the provided labels.
+
+    Args:
+      labels_filter: a dict of labels to find --> {'id': '123'}
+
+    Returns:
+      A call to __list_by_label with the proper instance_service_object
+    """
+
+    instance_service_object = self.gce_api().instances()
+    return self.__list_by_label(labels_filter, instance_service_object)
+
+  def list_disk_by_label(self, labels_filter=''):
+    """Lists Disks in a project with on of the provided labels.
+
+    Args:
+      labels_filter: a dict of labels to find --> {'id': '123'}
+
+    Returns:
+      A call to __list_by_label with the proper disk_service_object
+    """
+
+    disk_service_object = self.gce_api().disks()
+    return self.__list_by_label(labels_filter, disk_service_object)
+
+  def __list_by_label(self, labels_filter, service_object):
+    """List Disks or VMs in a project with one of the provided labels.
+
+    Args:
+      labels_filter:
+      service_object: Google Compute Engine (Disk | Instance) service object
+
+    Returns:
+      Dictionary with name and metadata(zone, labels) for each instance.
+      ex: {'instance-1': {'zone': 'us-central1-a', 'labels': {'id': '123'}}
+    """
+    resource_dict = dict()
+    filter_expression = ''
+    for key, value in labels_filter.items():
+      filter_expression = 'labels.' + key + '=' + value
+      request = service_object.aggregatedList(
+          project=self.project_id, filter=filter_expression)
```

Comment:
You're calling `aggregatedList` for every key/value pair in `label_filter`. That's OK, but maybe it would be interesting to `AND` the filters together if the user provides more than one filter?

Reply:
this will be more efficient yes, however the new questions is, should it be OR or AND or two different versions of each method or additional argument to select?

Reply:
I refactored the code to 4 new methods instead of two and the filter expression supplied to the API call will be either the anded or ored version of the input dict:
list_instance_by_labels_ored
list_disk_by_labels_ored
list_instance_by_labels_anded
list_disk_by_labels_anded